cpp/src/parquet/types.cc

```diff
@@ -96,57 +98,125 @@ bool PageCanUseChecksum(PageType::type pageType) {
+          reinterpret_cast<const uint8_t*>(val.data()), static_cast<int32_t>(val.size()));
+      if (!decimal_result.ok()) {
+        throw ParquetException("Failed to parse decimal value: ",
+                               decimal_result.status().message());
+      }
+      result << decimal_result.ValueUnsafe().ToString(scale);
+      break;
+    }
+    default:
+      throw ParquetException("Unsupported decimal type: ", TypeToString(parquet_type));
+  }
+
+  return result.str();
+}
+
+std::string FormatNonUTF8Value(::std::string_view val) {
+  if (val.empty()) {
+    return "";
+  }
+
   std::stringstream result;
+  result << "0x" << std::hex;
+  for (const auto& c : val) {
+    result << std::setw(2) << std::setfill('0')
+           << static_cast<int>(static_cast<unsigned char>(c));
+  }
+  return result.str();
+}
+
+}  // namespace
 
+std::string FormatStatValue(Type::type parquet_type, ::std::string_view val,
+                            const std::shared_ptr<const LogicalType>& logical_type) {
   const char* bytes = val.data();
   switch (parquet_type) {
     case Type::BOOLEAN: {
-      bool value{};
-      std::memcpy(&value, bytes, sizeof(bool));
-      result << value;
-      break;
+      return FormatNumericValue<bool>(val);
     }
     case Type::INT32: {
-      int32_t value{};
-      std::memcpy(&value, bytes, sizeof(int32_t));
-      result << value;
-      break;
+      if (logical_type != nullptr && logical_type->is_decimal()) {
+        return FormatDecimalValue(parquet_type, val, logical_type);
+      }
+      return FormatNumericValue<int32_t>(val);
     }
     case Type::INT64: {
-      int64_t value{};
-      std::memcpy(&value, bytes, sizeof(int64_t));
-      result << value;
-      break;
+      if (logical_type != nullptr && logical_type->is_decimal()) {
+        return FormatDecimalValue(parquet_type, val, logical_type);
+      }
+      return FormatNumericValue<int64_t>(val);
     }
     case Type::DOUBLE: {
-      double value{};
-      std::memcpy(&value, bytes, sizeof(double));
-      result << value;
-      break;
+      return FormatNumericValue<double>(val);
     }
     case Type::FLOAT: {
-      float value{};
-      std::memcpy(&value, bytes, sizeof(float));
-      result << value;
-      break;
+      return FormatNumericValue<float>(val);
     }
     case Type::INT96: {
+      std::stringstream result;
       std::array<int32_t, 3> values{};
       std::memcpy(values.data(), bytes, 3 * sizeof(int32_t));
       result << values[0] << " " << values[1] << " " << values[2];
-      break;
+      return result.str();
     }
     case Type::BYTE_ARRAY:
     case Type::FIXED_LEN_BYTE_ARRAY: {
-      result << val;
-      break;
+      if (logical_type != nullptr) {
+        if (logical_type->is_decimal()) {
+          return FormatDecimalValue(parquet_type, val, logical_type);
+        }
+        if (logical_type->is_string()) {
+          return std::string(val);
+        }
```

Comment:
What about fp16?

Reply:
Added. I think we are still missing format functions for other logical types like timestamp, geometry, etc. We can add them as needed.